crates/libs/windows/src/core/abi.rs

```diff
@@ -1,48 +1,73 @@
 use super::*;
 
+/// Describes how a type can be transfered across FFI boundaries
+///
+/// # Safety
+///
+/// * The associated type `Abi` must be safe to transfer over FFI boundaries (e.g., it must have a stable layout)
+/// * `from_abi` must be implemented if there are some in-memory representations of `Abi` that are not valid representations of `Self`
 #[doc(hidden)]
 pub unsafe trait Abi: Sized {
+    /// The type used to represent `Self` across FFI boundaries
     type Abi;
 
+    /// Converts an abi value to `Self` or fails
+    ///
     /// # Safety
+    ///
+    /// This function is safe to call if `abi` can safely be trivial transmuted to `Self`
     unsafe fn from_abi(abi: Self::Abi) -> Result<Self> {
         Ok(core::mem::transmute_copy(&abi))
     }
 
     /// # Safety
-    unsafe fn drop_param(_: &mut Param<Self>) {}
+    ///
+    /// This function is only safe to call inside of `_param`'s `Drop` implementation
+    /// as `_param` may contain freed memory inside of it after this function is run
+    unsafe fn drop_param(_param: &mut Param<Self>) {}
 }
 
+// SAFETY: raw pointers are FFI safe, `Abi` & `Self` are the same so thus they have the same in-memory representation, and
+// all representations of `Abi` are valid representations of `Self`.
 unsafe impl<T> Abi for *mut T {
     type Abi = Self;
 }
 
+// SAFETY: see the justification for `*mut T`
 unsafe impl<T> Abi for *const T {
     type Abi = Self;
 }
 
 unsafe impl<T: Interface> Abi for T {
     type Abi = RawPtr;
```

Comment:
Of the impls in this file, I think this is the only one that requires `from_abi` to be unsafe, since it erases the type of the interface. I wonder if we could hang on to the type information? For example, maybe something like this would work?

```rust
unsafe impl<T> Abi for NonNull<T> where Self: Interface {
    type Abi = *const T;
    ...
}
```

A [rough sketch](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cba771cd6dfcb73aa8a029136117652b) of it compiles, although I don't think this works with the rest of the codebase. The main reason is it looks like `Interface` objects are usually structs that wrap an `IUnknown`.

I guess another option is to not have this blanket impl and instead generate one for every interface type we generate. That's probably horrible for compile times though.

Also, I'm guessing most APIs take a void pointer rather than knowing the exact type, so we probably need the type-erased impl anyway.

Anyway, it's probably best to keep the trait and/or the `from_abi` method unsafe for now, but it seems like in theory we could factor things to keep type safety

Reply:
A big problem I have is not being to have more blanket impls - there's just so much boilerplate code that I'd love to get rid of. Short of derive macros, I'd love to find a way to have more default behavior inheritable. Even with derive macros, this would seem cleaner.

Reply:
Even if we have the interface type in `Abi` this would still have to be unsafe. Reading raw pointers is always unsafe since so many invariants are not upheld by the compiler, and it is up to the code that transforms a raw pointer to a more concrete Rust type to ensure those invariants have been upheld. For example:

*raw pointers are not type safe (i.e., casting raw pointers is "safe")*

```rust
let ptr: *const usize = &42usize as *const usize;
let ptr = ptr as *const IUknown::Vtable; // this line is wrong but not considered "unsafe"
// It is up to the next line to make sure `ptr` is well formed (which it is not)
unsafe { IUnknown::from_abi(ptr) }
```

Reply:
Ah, good point. The fact that casting between different unsafe pointer traits is safe means `from_abi` has to be unsafe.